File: leetcode75/js/1456-maximum-number-of-vowels-in-a-substring-of-given-length.js
var maxVowels = function(s, k) {
    const vowels = new Set(['a', 'e', 'i', 'o', 'u']);
    let maxCount = 0;
    let currentCount = 0;

    // Count vowels in the first window of size k
    for (let i = 0; i < k; i++) {
        if (vowels.has(s[i])) {
            currentCount++;
        }
    }
    maxCount = currentCount;

    // Slide the window over the string
    for (let i = k; i < s.length; i++) {
        if (vowels.has(s[i])) {
            currentCount++;
        }
        if (vowels.has(s[i - k])) {
            currentCount--;
        }
        maxCount = Math.max(maxCount, currentCount);
    }

    return maxCount;
};

var s = "abciiidef", k = 3
console.log(maxVowels(s, k)) // 3
var s = "aeiou", k = 2
console.log(maxVowels(s, k)) // 2
var s = "leetcode", k = 3
console.log(maxVowels(s, k)) // 2

File: leetcode75/go/1456-maximum-number-of-vowels-in-a-substring-of-given-length.go
package main

import (
	"fmt"
)

func main() {
	s := "abciiidef"
	k := 3
	result := maxVowels(s, k)
	fmt.Println(result) // Output: 3
}

func maxVowels(s string, k int) int {
	maxVowels := 0
	currentVowels := 0
	var vowels [128]bool
	vowels['a'], vowels['e'], vowels['i'], vowels['o'], vowels['u'] = true, true, true, true, true

	for i := 0; i < k; i++ {
		if vowels[s[i]] {
			currentVowels++
		}
	}

	maxVowels = currentVowels

	for i := k; i < len(s); i++ {
		if vowels[s[i]] {
			currentVowels++
		}
		if vowels[s[i-k]] {
			currentVowels--
		}
		if currentVowels > maxVowels {
			maxVowels = currentVowels
		}
	}

	return maxVowels
}
